File: framework/java/src/main/java/soda/unittest/job/JobSpec.java
package soda.unittest.job;

import java.lang.reflect.Method;

public class JobSpec {

	public Class<?> jobClass;
	
	public Method method;
	
	public Class<?> retClass;
	
	public Class<?>[] argClasses;
	
	public boolean validateByObject = false;
	
	public Validator<?> objectValidator;
	
	public Validator<?> serialValidator;
	
	public JobSpec(Class<?> jobClass, String method) throws NoSuchMethodException {
		this.jobClass = jobClass;
		this.method = findMethod(jobClass, method);
		retClass = this.method.getReturnType();
		argClasses = this.method.getParameterTypes();
	}
	
	private Method findMethod(Class<?> jobClass, String methodName) throws NoSuchMethodException {
		Method[] methods = jobClass.getMethods();
		for (Method m : methods) {
			if (m.getName().equals(methodName)) {
				return m;
			}
		}
		throw new NoSuchMethodException(methodName);
	}
	
}

File: framework/scala/src/main/scala/soda/scala/unittest/job/JobSpec.scala
package soda.scala.unittest.job

import java.lang.reflect.Method

class JobSpec(var jobClass: Class[_], val methodName: String) {

    val method: Method = findMethod(jobClass, methodName)

    var retClass: Class[_] = method.getReturnType()

    var argClasses: Array[Class[_]] = method.getParameterTypes()

    var validateByObject: Boolean = false

    var objectValidator: Validator[_] = new DefaultValidator()

    var serialValidator: Validator[_] = new DefaultValidator()

    private def findMethod(jobClass: Class[_], methodName: String): Method = {
        val mo = jobClass.getMethods().find(_.getName() == methodName)
        if (mo == None) {
            throw new NoSuchMethodException(methodName)
        }
        mo.get
    }

}
